File: leet/207/n_e.java
import java.util.List;
import java.util.Map;
import java.util.LinkedList;
import java.util.HashMap;

public class Solution {
    private boolean[] visited;
    private boolean[] onStack;
    private boolean hasCycle = false;
    private Map<Integer, List<Integer>> edges;
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        // Initialize graph edges
        edges = new HashMap<>();
        for (int i = 0; i < numCourses; i++)
            edges.put(i, new LinkedList<>());
        for (int i = 0; i < prerequisites.length; i++)
            edges.get(prerequisites[i][0]).add(prerequisites[i][1]);
        // Initialize markers
        visited = new boolean[numCourses];
        onStack = new boolean[numCourses];
        // Start depth-first iteration
        for (int i = 0; i < visited.length; i++)
            if (!hasCycle && !visited[i])
                dfs(i);
        return !hasCycle;
    }

    private void dfs(int source) {
        onStack[source] = true;
        visited[source] = true;
        for (int next : edges.get(source)) {
            if (hasCycle) return;
            else if (onStack[next]) {
                hasCycle = true;
                return;
            }
            else if (!visited[next]) dfs(next); 
        }
        onStack[source] = false;
    }
}

File: leet/207/n_e.py
import unittest

class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        self.edges = []
        self.marked = []
        self.onStack = []
        self.hasCycle = False
        for i in xrange(numCourses):
            self.marked.append(False)
            self.onStack.append(False)
            self.edges.append([])
        for pair in prerequisites:
            self.edges[pair[0]].append(pair[1])
        for i in xrange(numCourses):
            if not self.hasCycle and not self.marked[i]:
                self.dfs(i)
        return not self.hasCycle

    def dfs(self, source):
        self.marked[source] = True
        self.onStack[source] = True
        for next in self.edges[source]:
            if self.hasCycle:
                return
            elif self.onStack[next]:
                self.hasCycle = True
                return
            elif not self.marked[next]:
                self.dfs(next)
        self.onStack[source] = False


class TestCases(unittest.TestCase):
    solution = Solution()

    def test(self):
        self.assertEqual(self.solution.canFinish(1, []), True)
        self.assertEqual(self.solution.canFinish(2, [[1, 0]]), True)
        self.assertEqual(self.solution.canFinish(2, [[1, 0], [0, 1]]), False)

if __name__ == '__main__':
    unittest.main()
